android/src/main/java/com/learnium/RNDeviceInfo/RNInstallReferrerClient.java

```diff
@@ -2,84 +2,123 @@
 
 import android.content.SharedPreferences;
 import android.content.Context;
-import android.os.RemoteException;
 import android.util.Log;
 
-import com.android.installreferrer.api.InstallReferrerClient;
-import com.android.installreferrer.api.InstallReferrerStateListener;
-import com.android.installreferrer.api.ReferrerDetails;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
 
-public class RNInstallReferrerClient {
 
-  private SharedPreferences sharedPreferences;
-  private InstallReferrerClient mReferrerClient;
+public class RNInstallReferrerClient {
+  private static Class<?> InstallReferrerClientClazz;
+  private static Class<?> InstallReferrerStateListenerClazz;
+  private static Class<?> ReferrerDetailsClazz;
 
-  RNInstallReferrerClient(Context context) {
-    sharedPreferences = context.getSharedPreferences("react-native-device-info", Context.MODE_PRIVATE);
-    mReferrerClient = InstallReferrerClient.newBuilder(context).build();
+  static {
     try {
-      mReferrerClient.startConnection(installReferrerStateListener);
+      InstallReferrerClientClazz = Class.forName("com.android.installreferrer.api.InstallReferrerClient");
+      InstallReferrerStateListenerClazz = Class.forName("com.android.installreferrer.api.InstallReferrerStateListener");
+      ReferrerDetailsClazz = Class.forName("com.android.installreferrer.api.ReferrerDetails");
     } catch (Exception e) {
-      // This is almost always a PermissionException. Log it and move on
-      System.err.println("RNInstallReferrerClient exception. getInstallReferrer will be unavailable: " + e.getMessage());
-      e.printStackTrace(System.err);
+      System.err.println("RNInstallReferrerClient exception. 'installreferrer' APIs are unavailable.");
     }
   }
 
-  private String getInstallReferrer() {
+  private final SharedPreferences sharedPreferences;
+  private Object mReferrerClient;
+  private Object installReferrerStateListener;
+
+  // From InstallReferrerClient.InstallReferrerResponse
+  private static final int R_RESPONSE_OK                    = 0;
+  private static final int R_RESPONSE_SERVICE_UNAVAILABLE   = 1;
+  private static final int R_RESPONSE_FEATURE_NOT_SUPPORTED = 2;
+
+  RNInstallReferrerClient(Context context) {
+    sharedPreferences = context.getSharedPreferences("react-native-device-info", Context.MODE_PRIVATE);
+
+    if (InstallReferrerClientClazz == null || InstallReferrerStateListenerClazz == null || ReferrerDetailsClazz == null) {
+      return;
```

Comment:
Just in case, could you throw in an info log explaining "The install referrer library was not found, so install referrer functionality will not be available" in here? silently bailing out of documented functionality is the way to cut this functionality out, but it might violate principle of least surprise for some un-wary dev

Reply:
That log line is already done in the static block which sets those.

Reply:
Line 23